File: src/src_core/fmg/core/img/StaticImg.java
package fmg.core.img;

import java.util.function.Consumer;

import fmg.common.Color;
import fmg.common.geom.Bound;
import fmg.common.geom.Size;
import fmg.common.notyfier.NotifyPropertyChanged;

/**
 * Abstract, platform independent, image characteristics
 *
 * @param <T> the entity of image
 * @param <TImage> plaform specific image
 **/
public abstract class StaticImg<T, TImage> extends NotifyPropertyChanged implements AutoCloseable {

   public static Consumer<Runnable> DEFERR_INVOKER;

   public static final Color DefaultBkColor = new Color(0xFF, 0xFF, 0x8C, 0x00);
   public static final int DefaultImageSize = 100;

   protected StaticImg(T entity) {
      _entity = entity;
      _size = new Size(DefaultImageSize, DefaultImageSize);
      _padding = new Bound((int)(DefaultImageSize * 0.05)); // 5%
   }

   @SuppressWarnings("deprecation")
   protected <TI>boolean setProperty(TI storage, TI value, String propertyName) {
      return super.setProperty(value, propertyName);
   }

   public static final String PROPERTY_SIZE              = "Size";
   public static final String PROPERTY_PADDING           = "Padding";
   public static final String PROPERTY_ENTITY            = "Entity";
   public static final String PROPERTY_IMAGE             = "Image";
   public static final String PROPERTY_BACKGROUND_COLOR  = "BackgroundColor";
   public static final String PROPERTY_BORDER_COLOR      = "BorderColor";
   public static final String PROPERTY_BORDER_WIDTH      = "BorderWidth";
   public static final String PROPERTY_ROTATE_ANGLE      = "RotateAngle";
   public static final String PROPERTY_FOREGROUND_COLOR  = "ForegroundColor";

   private Size _size;
   /** width and height in pixel */
   public Size getSize() { return _size; }
   public void setSize(int widhtAndHeight) { setSize(new Size(widhtAndHeight, widhtAndHeight)) ; }
   public void setSize(Size value) {
      if (setProperty(_size, value, PROPERTY_SIZE)) {
         setImage(null);
         //invalidate();
      }
   }

   /** width image */
   public int getWidth() { return getSize().width; }
   /** height image */
   public int getHeight() { return getSize().height; }

   private Bound _padding;
   /** inside padding */
   public Bound getPadding() { return _padding; }
   public void setPadding(int bound) { setPadding(new Bound(bound)); }
   public void setPadding(Bound value) {
      if (value.getLeftAndRight() >= getWidth())
         throw new IllegalArgumentException("Padding size is very large. Should be less than Width.");
      if (value.getTopAndBottom() >= getHeight())
         throw new IllegalArgumentException("Padding size is very large. Should be less than Height.");
      if (setProperty(_padding, value, PROPERTY_PADDING)) {
         invalidate();
      }
   }

   public T _entity;
   public T getEntity() { return _entity; }
   public void setEntity(T value) {
      if (setProperty(_entity, value, PROPERTY_ENTITY))
         invalidate();
   }

   private enum EInvalidate {
      needRedraw,
      redrawing,
      redrawed
   }
   private EInvalidate _invalidate = EInvalidate.needRedraw;

   protected abstract TImage createImage();
   private TImage _image;
   public TImage getImage() {
      if (_image == null) {
         setImage(createImage());
         _invalidate = EInvalidate.needRedraw;
      }
      if (_invalidate == EInvalidate.needRedraw)
         draw();
      return _image;
   }
   protected void setImage(TImage value) {
      setProperty(_image, value, PROPERTY_IMAGE);
   }

   private Color _backgroundColor = DefaultBkColor;
   /** background fill color */
   public Color getBackgroundColor() { return _backgroundColor; }
   public void setBackgroundColor(Color value) {
      if (setProperty(_backgroundColor, value, PROPERTY_BACKGROUND_COLOR))
         invalidate();
   }

   private Color _borderColor = Color.Maroon.clone().darker(0.5);
   public Color getBorderColor() { return _borderColor; }
   public void setBorderColor(Color value) {
      if (setProperty(_borderColor, value, PROPERTY_BORDER_COLOR))
         invalidate();
   }

   private int _borderWidth = 3;
   public int getBorderWidth() { return _borderWidth; }
   public void setBorderWidth(int value) {
      if (setProperty(_borderWidth, value, PROPERTY_BORDER_WIDTH))
         invalidate();
   }

   private double _rotateAngle;
   /** 0° .. +360° */
   public double getRotateAngle() { return _rotateAngle; }
   public void setRotateAngle(double value) {
      if ((value > 360) || (value < 0)) {
         value %= 360;
         if (value < 0)
            value += 360;
      }
      if (setProperty(_rotateAngle, value, PROPERTY_ROTATE_ANGLE))
         invalidate();
   }

   private Color _foregroundColor = Color.Aqua;
   public Color getForegroundColor() { return _foregroundColor; }
   public void setForegroundColor(Color value) {
      if (setProperty(_foregroundColor, value, PROPERTY_FOREGROUND_COLOR))
         invalidate();
   }

   private boolean _deferredNotifications = true;
   public boolean isDeferredNotifications() { return _deferredNotifications; }
   public void setDeferredNotifications(boolean value) { _deferredNotifications = value; }

   protected void invalidate() {
      if (_invalidate == EInvalidate.redrawing)
         return;
      //if (_invalidate == EInvalidate.needRedraw)
      //   return;
      _invalidate = EInvalidate.needRedraw;
      onSelfPropertyChanged("Image");
   }

   private void draw() {
      drawBegin();
      drawBody();
      drawEnd();
   }

   protected void drawBegin() { _invalidate = EInvalidate.redrawing; }
   protected abstract void drawBody();
   protected void drawEnd() { _invalidate = EInvalidate.redrawed; }

   /** Deferr notifications */
   @Override
   protected void onSelfPropertyChanged(Object oldValue, Object newValue, String propertyName) {
      if (!isDeferredNotifications())
         super.onSelfPropertyChanged(oldValue, newValue, propertyName);
      else
         DEFERR_INVOKER.accept( () -> super.onSelfPropertyChanged(oldValue, newValue, propertyName) );
   }

   @Override
   public void close() { }

}

File: src/src_core/fmg/core/img/StaticImg.cs
using System;
using System.Linq;
using System.ComponentModel;
using System.Collections.Generic;
using fmg.common;
using fmg.common.geom;
using fmg.common.notyfier;

namespace fmg.core.img {

   public static class StaticImgConsts {

      public static Action<Action> DeferrInvoker;
      public static readonly Color DefaultBkColor = new Color(0xFF, 0xFF, 0x8C, 0x00);
      public static readonly Color DefaultForegroundColor = Color.LightSeaGreen;

   }

   /// <summary> Abstract, platform independent, image characteristics </summary>
   /// <typeparam name="TImage">plaform specific image</typeparam>
   public abstract class StaticImg<TImage> : NotifyPropertyChanged
      where TImage : class
   {
      public const int DefaultImageSize = 100;
      public const int DefaultPaddingInt = (int)(DefaultImageSize* 0.05); // 5%

      protected StaticImg() {
         _size = new Size(DefaultImageSize, DefaultImageSize);
         _padding = new Bound(DefaultPaddingInt);
      }

      private Size _size;
      /// <summary> width and height in pixel </summary>
      public Size Size {
         get { return _size; }
         set {
            var old = _size;
            if (SetProperty(ref _size, value)) {
               Image = null;
               //Invalidate();
            }
         }
      }

      private Bound _padding;
      /// <summary> inside padding </summary>
      public Bound Padding {
         get { return _padding; }
         set {
            if (value.LeftAndRight >= Size.Width)
               throw new ArgumentException("Padding size is very large. Should be less than Width.");
            if (value.TopAndBottom >= Size.Height)
               throw new ArgumentException("Padding size is very large. Should be less than Height.");
            if (SetProperty(ref _padding, value)) {
               Invalidate();
            }
         }
      }
      public int PaddingInt {
         set { Padding = new Bound(value); }
      }

      private enum EInvalidate {
         NeedRedraw,
         Redrawing,
         Redrawed
      }
      private EInvalidate _invalidate = EInvalidate.NeedRedraw;
      protected abstract TImage CreateImage();
      private TImage _image;
      public TImage Image {
         get {
            //LoggerSimple.Put("getImage: {0}", Entity);
            if (_image == null) {
               Image = CreateImage();
               _invalidate = EInvalidate.NeedRedraw;
            }
            if (_invalidate == EInvalidate.NeedRedraw)
               Draw();
            return _image;
         }
         protected set {
            SetProperty(ref _image, value);
         }
      }

      private Color _backgroundColor = StaticImgConsts.DefaultBkColor;
      /// <summary> background fill color </summary>
      public Color BackgroundColor {
         get { return _backgroundColor; }
         set {
            if (SetProperty(ref _backgroundColor, value))
               Invalidate();
         }
      }

      private Color _borderColor = Color.Maroon.Darker(0.5);
      public Color BorderColor {
         get { return _borderColor; }
         set {
            if (SetProperty(ref _borderColor, value))
               Invalidate();
         }
      }

      private int _borderWidth = 3;
      public int BorderWidth {
         get { return _borderWidth; }
         set {
            if (SetProperty(ref _borderWidth, value))
               Invalidate();
         }
      }

      private double _rotateAngle;
      /// <summary> 0° .. +360° </summary>
      public double RotateAngle {
         get { return _rotateAngle; }
         set {
            value = FixAngle(value);
            if (SetProperty(ref _rotateAngle, value))
               Invalidate();
         }
      }

      /// <summary> to diapason (0° .. +360°] </summary>
      protected static double FixAngle(double value) {
         return (value >= 360)
              ?               (value % 360)
              : (value < 0)
                 ?            (value % 360) + 360
                 :             value;
      }


      private Color _foregroundColor = StaticImgConsts.DefaultForegroundColor;
      public Color ForegroundColor {
         get { return _foregroundColor; }
         set {
            if (SetProperty(ref _foregroundColor, value))
               Invalidate();
         }
      }

      public bool DeferredNotifications { get; set; } = true;
      private readonly Dictionary<string, PropertyChangedEventArgs> DeferredNotificationsMap = new Dictionary<string, PropertyChangedEventArgs>();

      protected void Invalidate() {
         if (_invalidate == EInvalidate.Redrawing)
            return;
         //if (_invalidate == EInvalidate.NeedRedraw)
         //   return;
         _invalidate = EInvalidate.NeedRedraw;
         OnSelfPropertyChanged(nameof(this.Image));
      }

      private void Draw() {
         //LoggerSimple.Put("> DrawBegin: {0}", Entity);
         DrawBegin();
         DrawBody();
         DrawEnd();
         //LoggerSimple.Put("< DrawEnd: {0}", Entity);
      }

      protected virtual void DrawBegin() { _invalidate = EInvalidate.Redrawing; }
      protected abstract void DrawBody();
      protected virtual void DrawEnd() { _invalidate = EInvalidate.Redrawed; }

      protected override void OnSelfPropertyChanged(PropertyChangedEventArgs ev) {
         if (!DeferredNotifications) {
            base.OnSelfPropertyChanged(ev);
            //OnPropertyChangingAfter(true, ev);
         } else {
            bool first = DeferredNotificationsMap.Any();
            if (!DeferredNotificationsMap.ContainsKey(ev.PropertyName))
               DeferredNotificationsMap.Add(ev.PropertyName, ev);
            else
               DeferredNotificationsMap[ev.PropertyName] = ev;
            if (first)
               StaticImgConsts.DeferrInvoker(() => {
                  foreach (PropertyChangedEventArgs item in DeferredNotificationsMap.Values.ToList()) {
                     base.OnSelfPropertyChanged(item);
                     //OnSelfPropertyChangedAfter(false, item);
                  }
                  DeferredNotificationsMap.Clear();
               });
         }
      }

      //protected abstract void OnSelfPropertyChangedAfter(bool sync, object sender, PropertyChangedEventArgs ev);

   }

}
